### ISSUE TITLE
JSON RPC WebSocket Methods Missing from Experimental Lib

### ISSUE THREAD

**Issue by buffalojoec** [solana-labs/solana-web3.js#1505]
- [ ] [`accountSubscribe`](https://docs.solana.com/api/websocket#accountsubscribe)
- [ ] [`accountUnsubscribe`](https://docs.solana.com/api/websocket#accountunsubscribe)
- [ ] [`blockSubscribe`](https://docs.solana.com/api/websocket#blocksubscribe)
- [ ] [`blockUnsubscribe`](https://docs.solana.com/api/websocket#blockunsubscribe)
- [ ] [`logsSubscribe`](https://docs.solana.com/api/websocket#logssubscribe)
- [ ] [`logsUnsubscribe`](https://docs.solana.com/api/websocket#logsunsubscribe)
- [ ] [`programSubscribe`](https://docs.solana.com/api/websocket#programsubscribe)
- [ ] [`programUnsubscribe`](https://docs.solana.com/api/websocket#programunsubscribe)
- [ ] [`signatureSubscribe`](https://docs.solana.com/api/websocket#signaturesubscribe)
- [ ] [`signatureUnsubscribe`](https://docs.solana.com/api/websocket#signatureunsubscribe)
- [ ] [`slotSubscribe`](https://docs.solana.com/api/websocket#slotsubscribe)
- [ ] [`slotUnsubscribe`](https://docs.solana.com/api/websocket#slotunsubscribe)
- [ ] [`slotsUpdatesSubscribe`](https://docs.solana.com/api/websocket#slotsupdatessubscribe)
- [ ] [`slotsUpdatesUnsubscribe`](https://docs.solana.com/api/websocket#slotsupdatesunsubscribe)
- [ ] [`voteSubscribe`](https://docs.solana.com/api/websocket#votesubscribe)
- [ ] [`voteUnsubscribe`](https://docs.solana.com/api/websocket#voteunsubscribe)

**Comment by steveluscher** [comment]
So here's what I'm thinking, as a gameplan.

# Add `transports/websocket/` to `@solana/rpc-transport`, or maybe create a `@solana/rpc-subscriptions-transport` package

This will be the simplest possible wrapper around `WebSocket`. It should have, as a design goal, ‘to actually work.’

An anti-goal will be to build it into the regular JSON RPC implementation. This is to prevent forcing devs to bundle the subscriptions code even if they don't use it.

This work will produce the `IRpcWebsocketTransport` type and a concrete implementation of it.

# Create typespecs for each subscription in `@solana/rpc-core`

These will be slightly different than the request-response ones in that they'll:

* Take in an `AbortSignal` that can be used to unsubscribe
* Return an `AsyncIterator` that yields subscription notifications until you unsubscribe or the transport fails

```ts
interface AccountSubscribeApi {
    accountNotification(
        address: Base58EncodedAddress,
        config?: AccountSubscribeApiConfig,
    ): AsyncIterator<AccountNotification>
}
```

The choice of `AsyncIterator` here – instead of `Observable` or `EventEmitter` – is owing to the fact that `AsyncIterator` is supported by the platform whereas the other two require polyfills / userspace implementations.

# Glue the typespec and the transport together in `@solana/rpc-transport`

Proxy calls that conform to the typespec to the websocket transport in the same way that we do with the regular RPC now. Coalesce identical calls behind the same subscription.

```ts
// Subscribe for updates to an account.
const accountSubscriptionController = new AbortController();
const accountNotifications = await rpc
    .accountNotification('96pjpihod3x81wLaQmcwmznYKpJQzbQzZtLMXa8y6Lqv')
    .subscribe({unsubscribeSignal: accountSubscriptionController.signal});
// Spawn a loop that listens for notifications indefinitely.
(async () => {
    try {
        for await (const notification of accountNotifications) {
            console.log('account was updated', notification);
        }
    } catch (e) {
        // Handle a fatal in the transport (eg. implement retry logic)
        console.error('subscription died', e.message);
    }
})();
// Wait for 10 seconds then unsubscribe from notifications.
setTimeout(() => {
    accountSubscriptionController.abort();
}, 10_000);
```

_Super_ notable here is that we simply throw when the transport goes down for any reason rather than to retry, giving the application developer the opportunity to decide how to handle it. This is extremely important because:

* they might need to do something to ‘catch up’ on all the notifications they _missed_ between the time the transport went down, and the time the subscription was re-established.
* they might like to show some UI to indicate that subscriptions is dead
* they might like to define their _own_ retry logic/cadence depending on their relationship with their RPC provider

Open questions:
1. Should we _suspend_ execution when calling `subscribe()` until the websocket transport is set up, or make socket setup just another thing that the code consuming the `AsyncIterator` waits for?
2. Should we move `json-rpc.ts` out of `@solana/rpc-transport` and into `@solana/rpc` or something?

# Optionally explore meta tooling

1. We could create (or encourage others to create) wrappers around the new transport that do things like auto-retry
2. We could create (or encourage others to create) React hooks like `useSolanaSubscription()` that do automatic setup and teardown when components unmount/remount.

**Comment by buffalojoec** [comment]
@steveluscher this seems overall like a really solid plan.

Regardless of whether or not we go with a separate package, do you think it makes sense to rename some of the existing transport stuff for request-response? Considering we're dubbing everything here with "subscription", is there some nomenclature we should use for the other transport, like "request"? ie:

```text
@solana/rpc-request-transport
@solana/rpc-subscription-transport
```

> 1. Should we _suspend_ execution when calling `subscribe()` until the websocket transport is set up, or make socket setup just another thing that the code consuming the `AsyncIterator` waits for?

I guess I'm not really sure what the trade-off is here. Intuitively I think the latter makes more sense; don't begin execution until everything is set up. Is there some more insight you can give to this comparison?

> 2. Should we move `json-rpc.ts` out of `@solana/rpc-transport` and into `@solana/rpc` or something?

This kind of ties into my question above. This proxy configuration is for request-response, right? So we could just align it with the request-response track of the RPC and insert the subscription model alongside it, then bring both into scope using the main `@solana/web3.js` (experimental) lib, but still as separate transports to avoid the unintentional bundling of both.

> 1. We could create (or encourage others to create) wrappers around the new transport that do things like auto-retry
> 2. We could create (or encourage others to create) React hooks like `useSolanaSubscription()` that do automatic setup and teardown when components unmount/remount.

These are both cool ideas, but I would lean more towards seeing the community lay down the meta tooling they collectively decide they need.

What if instead of trying to anticipate areas that developers will likely build on top of web3next, we instead create a means for the submission, discussion, collaborative development, and broadcasting of releases for any such tools?

TL/DR, we help foster creative add-ons for web3next built by community members and build some sort of "plugin marketplace" for web3next extensions. The most popular add-ons could be easy layups for public goods funding (Cubik, etc.).

**Comment by steveluscher** [comment]
> …do you think it makes sense to rename some of the existing transport stuff for request-response?

Maybe! I'm _really_ not that worried about it. We can get it all working end-to-end and then thrash everything apart later.

> …don't begin execution until everything is set up. Is there some more insight you can give to this comparison?

This is the thing I'm struggling with most now. The struggle is between being able to use the `AsyncIterator` API to poll for messages without needing to know if the connection is set up, versus waiting until the connection is set up but _potentially_ missing a message. The latter would be devastating in the case of something like `signatureSubscribe`.

```ts
// Imagine this...
const signatureNotifications = rpc
  .signatureNotifications(signature)
  .subscribe({abortSignal});

// Is `signatureNotifications` a promise for an `AsyncIterator`?
const notificationsIterator = await signatureNotifications;
for await (const notification in notificationsIterator) {
  console.log('got a signature notification', notification);
}

// Or is `signatureNotifications` the iterator itself?
for await (const notification in signatureNotifications) {
  console.log('got a signature notification', notification);
}

// When did the websocket connection actually get set up?
const notificationsIterator = await signatureNotifications; // <--- HERE?
for await (const notification in ...) // <--- OR HERE?

// If it gets set up when you `await` the result of `subscribe()`, do you run into this race condition?
const notificationsIterator = await signatureNotifications; // <--- SUBSCRIPTION MADE
// ...time passes...
for await (const notification in ...) // <--- DID YOU MISS THE NOTIFICATION?
```

> … but still as separate transports to avoid the unintentional bundling of both.

Mmmm. I will have to think about whether there's a `web3.js` user that only wants subscriptions, or only wants request-response, and figure out how they'll be able to get one without the other.

> …I would lean more towards seeing the community lay down the meta tooling they collectively decide they need.

Same.

**Comment by steveluscher** [comment]
Pretty sure we knocked all of these out 🎉
